commit a45b2ea11c15f35c36330ff27cb45854a29c2e2c
Author: Chris Wilson <chris@chris-wilson.co.uk>
Date:   Thu Jul 17 14:40:51 2014 +0100

    sna: Handle rotated slaved scanouts
    
    As per the usual handling of rotated scanouts with unsupported
    rotations, we stage the drawing onto a backbuffer then rotate onto the
    crtc later. The difference here is that we must read the contents from
    the master pixmap rather than our own screen pixmap.
    
    Spotted was looking at slaved scanouts mistakenly setting the CRTC
    transformed flag.
    
    Reported-by: Tomas Pruzina <pruzinat@gmail.com>
    Bugzilla: https://bugs.freedesktop.org/show_bug.cgi?id=81383
    Signed-off-by: Chris Wilson <chris@chris-wilson.co.uk>

--- a/src/sna/sna_display.c
+++ b/src/sna/sna_display.c
@@ -112,7 +112,8 @@
 struct sna_crtc {
 	struct drm_mode_modeinfo kmode;
 	int dpms_mode;
-	PixmapPtr scanout_pixmap;
+	PixmapPtr slave_pixmap;
+	DamagePtr slave_damage;
 	struct kgem_bo *bo, *shadow_bo;
 	uint32_t cursor;
 	bool shadow;
@@ -988,7 +989,7 @@
 	VG_CLEAR(arg);
 	arg.crtc_id = sna_crtc->id;
 	arg.fb_id = fb_id(sna_crtc->bo);
-	if (sna_crtc->transform) {
+	if (sna_crtc->transform || sna_crtc->slave_pixmap) {
 		arg.x = 0;
 		arg.y = 0;
 	} else {
@@ -1161,7 +1162,23 @@
 	assert(sna->mode.shadow_active == 0);
 }
 
-static bool sna_crtc_enable_shadow(struct sna *sna, struct sna_crtc *crtc)
+static void sna_crtc_slave_damage(DamagePtr damage, RegionPtr region, void *closure)
+{
+	xf86CrtcPtr base = closure;
+	struct sna_crtc *crtc = to_sna_crtc(base);
+	struct sna *sna = to_sna(base->scrn);
+	RegionPtr scr;
+
+	assert(crtc->slave_damage == damage);
+	assert(sna->mode.shadow_damage);
+
+	RegionTranslate(region, base->x, base->y);
+	scr = DamageRegion(sna->mode.shadow_damage);
+	RegionUnion(scr, scr, region);
+	RegionTranslate(region, -base->x, -base->y);
+}
+
+static bool sna_crtc_enable_shadow(struct sna *sna, struct sna_crtc *crtc, xf86CrtcPtr base)
 {
 	if (crtc->shadow) {
 		assert(sna->mode.shadow_damage && sna->mode.shadow_active);
@@ -1177,6 +1194,21 @@
 		assert(sna->mode.shadow == NULL);
 	}
 
+	if (crtc->slave_pixmap) {
+		assert(crtc->slave_damage == NULL);
+
+		crtc->slave_damage = DamageCreate(sna_crtc_slave_damage, NULL,
+						  DamageReportRawRegion, TRUE,
+						  sna->scrn->pScreen, base);
+		if (crtc->slave_damage == NULL) {
+			if (!--sna->mode.shadow_active)
+				sna_mode_disable_shadow(sna);
+			return false;
+		}
+
+		DamageRegister(&crtc->slave_pixmap->drawable, crtc->slave_damage);
+	}
+
 	crtc->shadow = true;
 	sna->mode.shadow_active++;
 	return true;
@@ -1191,14 +1223,21 @@
 	DBG(("%s: disabling for crtc %d\n", __FUNCTION__, crtc->id));
 	assert(sna->mode.shadow_active > 0);
 
-	if (!--sna->mode.shadow_active)
-		sna_mode_disable_shadow(sna);
+	if (crtc->slave_damage) {
+		assert(crtc->slave_pixmap);
+		DamageUnregister(&crtc->slave_scanout->drawable, crtc->slave_damage);
+		DamageDestroy(crtc->slave_damage);
+		crtc->slave_damage = NULL;
+	}
 
 	if (crtc->shadow_bo) {
 		kgem_bo_destroy(&sna->kgem, crtc->shadow_bo);
 		crtc->shadow_bo = NULL;
 	}
 
+	if (!--sna->mode.shadow_active)
+		sna_mode_disable_shadow(sna);
+
 	crtc->shadow = false;
 }
 
@@ -1504,25 +1543,11 @@
 	sna_crtc->transform = false;
 	sna_crtc->rotation = RR_Rotate_0;
 
-	if (sna_crtc->scanout_pixmap) {
-		DBG(("%s: attaching to scanout pixmap\n", __FUNCTION__));
-
-		bo = sna_pixmap_pin(sna_crtc->scanout_pixmap, PIN_SCANOUT);
-		if (bo == NULL)
-			return NULL;
-
-		if (!get_fb(sna, bo,
-			    sna_crtc->scanout_pixmap->drawable.width,
-			    sna_crtc->scanout_pixmap->drawable.height))
-			return NULL;
-
-		sna_crtc->transform = true;
-		return kgem_bo_reference(bo);
-	} else if (use_shadow(sna, crtc)) {
+	if (use_shadow(sna, crtc)) {
 		unsigned long tiled_limit;
 		int tiling;
 
-		if (!sna_crtc_enable_shadow(sna, sna_crtc))
+		if (!sna_crtc_enable_shadow(sna, sna_crtc, crtc))
 			return NULL;
 
 		DBG(("%s: attaching to per-crtc pixmap %dx%d\n",
@@ -1557,7 +1582,7 @@
 	} else if (sna->flags & SNA_TEAR_FREE) {
 		DBG(("%s: tear-free updates requested\n", __FUNCTION__));
 
-		if (!sna_crtc_enable_shadow(sna, sna_crtc))
+		if (!sna_crtc_enable_shadow(sna, sna_crtc, crtc))
 			return NULL;
 
 		DBG(("%s: attaching to single shadow pixmap\n", __FUNCTION__));
@@ -1597,21 +1622,30 @@
 
 		return kgem_bo_reference(sna->mode.shadow);
 	} else {
-		unsigned rotation;
-
-		DBG(("%s: attaching to framebuffer\n", __FUNCTION__));
 		sna_crtc_disable_shadow(sna, sna_crtc);
-		bo = sna_pixmap_pin(sna->front, PIN_SCANOUT);
-		if (bo == NULL)
-			return NULL;
 
-		if (!get_fb(sna, bo, scrn->virtualX, scrn->virtualY))
-			return NULL;
+		if (sna_crtc->slave_pixmap) {
+			DBG(("%s: attaching to scanout pixmap\n", __FUNCTION__));
+			bo = sna_pixmap_pin(sna_crtc->slave_pixmap, PIN_SCANOUT);
+			if (bo == NULL)
+				return NULL;
+
+			if (!get_fb(sna, bo,
+				    sna_crtc->slave_pixmap->drawable.width,
+				    sna_crtc->slave_pixmap->drawable.height))
+				return NULL;
+		} else {
+			DBG(("%s: attaching to framebuffer\n", __FUNCTION__));
+			bo = sna_pixmap_pin(sna->front, PIN_SCANOUT);
+			if (bo == NULL)
+				return NULL;
 
-		assert(!sna_crtc->shadow);
-		rotation = rotation_reduce(&sna_crtc->primary, crtc->rotation);
-		assert(sna_crtc->primary.rotation.supported & rotation);
-		sna_crtc->rotation = rotation;
+			if (!get_fb(sna, bo, scrn->virtualX, scrn->virtualY))
+				return NULL;
+		}
+
+		sna_crtc->rotation = rotation_reduce(&sna_crtc->primary, crtc->rotation);
+		assert(sna_crtc->primary.rotation.supported & sna_crtc->rotation);
 		return kgem_bo_reference(bo);
 	}
 }
@@ -1984,10 +2018,19 @@
 static Bool
 sna_crtc_set_scanout_pixmap(xf86CrtcPtr crtc, PixmapPtr pixmap)
 {
+	struct sna_crtc *sna_crtc = to_sna_crtc(crtc);
+
+	if (sna_crtc == NULL)
+		return FALSE;
+
 	DBG(("%s: CRTC:%d, pipe=%d setting scanout pixmap=%ld\n",
-	     __FUNCTION__,to_sna_crtc(crtc)->id, to_sna_crtc(crtc)->pipe,
+	     __FUNCTION__, sna_crtc->id,  sna_crtc->pipe,
 	     pixmap ? pixmap->drawable.serialNumber : 0));
-	to_sna_crtc(crtc)->scanout_pixmap = pixmap;
+
+	if (pixmap == NULL)
+		sna_crtc_disable(crtc);
+	sna_crtc->slave_pixmap = pixmap;
+
 	return TRUE;
 }
 #endif
@@ -4547,11 +4590,26 @@
 		box->y2 = crtc->mode.VDisplay;
 }
 
+inline static DrawablePtr crtc_source(xf86CrtcPtr crtc, int16_t *sx, int16_t *sy)
+{
+	struct sna_crtc *sna_crtc = to_sna_crtc(crtc);
+	if (sna_crtc->slave_pixmap) {
+		*sx = -crtc->x;
+		*sy = -crtc->y;
+		return &sna_crtc->slave_pixmap->drawable;
+	} else {
+		*sx = *sy = 0;
+		return &to_sna(crtc->scrn)->front->drawable;
+	}
+}
+
 static void
 sna_crtc_redisplay__fallback(xf86CrtcPtr crtc, RegionPtr region, struct kgem_bo *bo)
 {
+	int16_t sx, sy;
 	struct sna *sna = to_sna(crtc->scrn);
 	ScreenPtr screen = sna->scrn->pScreen;
+	DrawablePtr draw = crtc_source(crtc, &sx, &sy);
 	PictFormatPtr format;
 	PicturePtr src, dst;
 	PixmapPtr pixmap;
@@ -4565,25 +4623,25 @@
 		return;
 
 	pixmap = sna_pixmap_create_unattached(screen,
-					      0, 0, sna->front->drawable.depth);
+					      0, 0, draw->depth);
 	if (pixmap == NullPixmap)
 		return;
 
 	if (!screen->ModifyPixmapHeader(pixmap,
 					crtc->mode.HDisplay,
 					crtc->mode.VDisplay,
-					sna->front->drawable.depth,
-					sna->front->drawable.bitsPerPixel,
+					draw->depth,
+					draw->bitsPerPixel,
 					bo->pitch, ptr))
 		goto free_pixmap;
 
-	error = sna_render_format_for_depth(sna->front->drawable.depth);
+	error = sna_render_format_for_depth(draw->depth);
 	format = PictureMatchFormat(screen,
 				    PIXMAN_FORMAT_DEPTH(error), error);
 	if (format == NULL) {
 		DBG(("%s: can't find format for depth=%d [%08x]\n",
-		     __FUNCTION__, sna->front->drawable.depth,
-		     (int)sna_render_format_for_depth(sna->front->drawable.depth)));
+		     __FUNCTION__, draw->depth,
+		     (int)sna_render_format_for_depth(draw->depth)));
 		goto free_pixmap;
 	}
 
@@ -4622,7 +4680,7 @@
 			     box.x1, box.y1, box.x2, box.y2));
 
 			fbComposite(PictOpSrc, src, NULL, dst,
-				    box.x1, box.y1,
+				    box.x1 + sx, box.y1 + sy,
 				    0, 0,
 				    box.x1, box.y1,
 				    box.x2 - box.x1, box.y2 - box.y1);
@@ -4640,8 +4698,10 @@
 static void
 sna_crtc_redisplay__composite(xf86CrtcPtr crtc, RegionPtr region, struct kgem_bo *bo)
 {
+	int16_t sx, sy;
 	struct sna *sna = to_sna(crtc->scrn);
 	ScreenPtr screen = crtc->scrn->pScreen;
+	DrawablePtr draw = crtc_source(crtc, &sx, &sy);
 	struct sna_composite_op tmp;
 	PictFormatPtr format;
 	PicturePtr src, dst;
@@ -4652,32 +4712,32 @@
 	DBG(("%s: compositing transformed damage boxes\n", __FUNCTION__));
 
 	pixmap = sna_pixmap_create_unattached(screen,
-					      0, 0, sna->front->drawable.depth);
+					      0, 0, draw->depth);
 	if (pixmap == NullPixmap)
 		return;
 
 	if (!screen->ModifyPixmapHeader(pixmap,
 					crtc->mode.HDisplay,
 					crtc->mode.VDisplay,
-					sna->front->drawable.depth,
-					sna->front->drawable.bitsPerPixel,
+					draw->depth,
+					draw->bitsPerPixel,
 					bo->pitch, NULL))
 		goto free_pixmap;
 
 	if (!sna_pixmap_attach_to_bo(pixmap, bo))
 		goto free_pixmap;
 
-	error = sna_render_format_for_depth(sna->front->drawable.depth);
+	error = sna_render_format_for_depth(draw->depth);
 	format = PictureMatchFormat(screen,
 				    PIXMAN_FORMAT_DEPTH(error), error);
 	if (format == NULL) {
 		DBG(("%s: can't find format for depth=%d [%08x]\n",
-		     __FUNCTION__, sna->front->drawable.depth,
-		     (int)sna_render_format_for_depth(sna->front->drawable.depth)));
+		     __FUNCTION__, draw->depth,
+		     (int)sna_render_format_for_depth(draw->depth)));
 		goto free_pixmap;
 	}
 
-	src = CreatePicture(None, &sna->front->drawable, format,
+	src = CreatePicture(None, draw, format,
 			    0, NULL, serverClient, &error);
 	if (!src)
 		goto free_pixmap;
@@ -4700,7 +4760,7 @@
 
 	if (!sna->render.composite(sna,
 				   PictOpSrc, src, NULL, dst,
-				   0, 0,
+				   sx, sy,
 				   0, 0,
 				   0, 0,
 				   crtc->mode.HDisplay, crtc->mode.VDisplay,
@@ -4738,10 +4798,11 @@
 static void
 sna_crtc_redisplay(xf86CrtcPtr crtc, RegionPtr region)
 {
+	int16_t tx, ty, sx, sy;
 	struct sna *sna = to_sna(crtc->scrn);
 	struct sna_crtc *sna_crtc = to_sna_crtc(crtc);
-	struct sna_pixmap *priv = sna_pixmap(sna->front);
-	int16_t tx, ty;
+	DrawablePtr draw = crtc_source(crtc, &sx, &sy);
+	struct sna_pixmap *priv = sna_pixmap((PixmapPtr)draw);
 
 	DBG(("%s: crtc %d [pipe=%d], damage (%d, %d), (%d, %d) x %ld\n",
 	     __FUNCTION__, sna_crtc->id, sna_crtc->pipe,
@@ -4756,7 +4817,7 @@
 
 		RegionTranslate(region, -crtc->bounds.x1, -crtc->bounds.y1);
 		sna_blt_fill_boxes(sna, GXcopy,
-				   sna_crtc->bo, sna->front->drawable.bitsPerPixel,
+				   sna_crtc->bo, draw->bitsPerPixel,
 				   priv->clear_color,
 				   REGION_RECTS(region), REGION_NUM_RECTS(region));
 		return;
@@ -4771,11 +4832,11 @@
 
 		tmp.drawable.width = crtc->mode.HDisplay;
 		tmp.drawable.height = crtc->mode.VDisplay;
-		tmp.drawable.depth = sna->front->drawable.depth;
-		tmp.drawable.bitsPerPixel = sna->front->drawable.bitsPerPixel;
+		tmp.drawable.depth = draw->depth;
+		tmp.drawable.bitsPerPixel = draw->bitsPerPixel;
 
 		if (sna->render.copy_boxes(sna, GXcopy,
-					   sna->front, priv->gpu_bo, 0, 0,
+					   container_of(draw, PixmapRec, drawable), priv->gpu_bo, sx, sy,
 					   &tmp, sna_crtc->bo, -tx, -ty,
 					   REGION_RECTS(region), REGION_NUM_RECTS(region), 0))
 			return;
@@ -4872,6 +4933,9 @@
 			if (RegionNotEmpty(&damage))
 				sna_crtc_redisplay__fallback(crtc, &damage, sna_crtc->bo);
 			RegionUninit(&damage);
+
+			if (sna_crtc->slave_damage)
+				DamageEmpty(sna_crtc->slave_damage);
 		}
 
 		RegionEmpty(region);
@@ -4992,6 +5056,9 @@
 			}
 		}
 		RegionUninit(&damage);
+
+		if (sna_crtc->slave_damage)
+			DamageEmpty(sna_crtc->slave_damage);
 	}
 
 	if (sna->mode.shadow) {
--- a/src/sna/sna_driver.c
+++ b/src/sna/sna_driver.c
@@ -593,7 +593,8 @@
 	DBG(("%s: triple buffer? %s\n", __FUNCTION__, sna->flags & SNA_TRIPLE_BUFFER ? "enabled" : "disabled"));
 
 	if ((sna->flags & (SNA_NO_VSYNC | SNA_NO_FLIP)) == 0 &&
-	    xf86ReturnOptValBool(sna->Options, OPTION_TEAR_FREE, FALSE))
+	    xf86ReturnOptValBool(sna->Options, OPTION_TEAR_FREE, FALSE) &&
+	    !sna->scrn->is_gpu)
 		sna->flags |= SNA_TEAR_FREE;
 
 	if (xf86ReturnOptValBool(sna->Options, OPTION_CRTC_PIXMAPS, FALSE))
